src/EditorFeatures/Core/Implementation/Structure/StructureTag.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.Structure;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.Structure
+{
+    internal sealed class StructureTag : IStructureTag
+    {
+        private readonly AbstractStructureTaggerProvider _tagProvider;
+
+        public StructureTag(AbstractStructureTaggerProvider tagProvider, BlockSpan blockSpan, ITextSnapshot snapshot)
+        {
+            Snapshot = snapshot;
+            OutliningSpan = blockSpan.TextSpan.ToSpan();
+            Type = ConvertType(blockSpan.Type);
+            IsCollapsible = blockSpan.IsCollapsible;
+            IsDefaultCollapsed = blockSpan.IsDefaultCollapsed;
+            IsImplementation = blockSpan.AutoCollapse;
+
+            // The HeaderSpan is what is used for drawing the guidelines and also what is shown if
+            // you mouse over a guideline. We will use the text from the hint start to the collapsing
+            // start; in the case this spans mutiple lines the editor will clip it for us and suffix an
+            // ellipsis at the end.
+            var headerSpan = Span.FromBounds(blockSpan.HintSpan.Start, blockSpan.TextSpan.Start);
+
+            if (headerSpan.Length > 0)
+            {
+                HeaderSpan = headerSpan;
+            }
+
+            CollapsedText = blockSpan.BannerText;
+            CollapsedHintFormSpan = blockSpan.HintSpan.ToSpan();
+            _tagProvider = tagProvider;
+        }
+
+        public ITextSnapshot Snapshot { get; }
+        public Span? OutliningSpan { get; }
+        public Span? HeaderSpan { get; }
+        public Span? GuideLineSpan => null;
+        public int? GuideLineHorizontalAnchorPoint => null;
+        public string Type { get; }
+        public bool IsCollapsible { get; }
+        public bool IsDefaultCollapsed { get; }
+        public bool IsImplementation { get; }
```

Comment:
1. props above constructor.
2. is IStructureTag a VS concept?  do they not have an impl we can use here?

Reply:
@CyrusNajmabadi: the interface is, but I don't think there's a useable implementation. Or at least since we override some methods to compute stuff on the fly. @gundermanc?

Reply:
There's a concrete implementation, [StructureTag](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.text.tagging.structuretag?view=visualstudiosdk-2019#:~:text=StructureTag(ITextSnapshot,%20Nullable%3CSpan%3E,%20Nullable%3CSpan%3E,%20Nullable%3CSpan%3E,%20Nullable%3CInt32%3E,%20String,%20Boolean,%20Boolean,,Object,%20Object)%20Constructs%20an%20instance%20of%20the%20IStructureTag.).

IIRC almost all of the members are virtual

Reply:
@CyrusNajmabadi Also as far as ordering, in this case the props are all the implementations of the interface, so was keeping that together with the methods below.

Reply:
Oh right, I didn't use that because the properties can't be assigned to in the constructor. Either you'd have to override (which defeats the point) or you have to jam all the logic into the base initializer which is prohibitive in some of the cases here.

Now I think it'd be good if the logic we have in StructureTag didn't exist and moved to other parts of the system, but alas this is also somewhat load bearing in terms of F# and TypeScript also pass through this so I'm somewhat constrained without expanding this PR to also involve them.